### PR TITLE
New Connector: spark-elastic

### LINKED ISSUE #1675: Provide new connector: spark-elastic
- [x] Running Elastic connector (runnable in UI)
- [x] Create workspaces
- [ ] Integration tests 100% passing
- [x] Spark cluster configurtion (memory, executors, etc)
- [ ] Deal with illegal chars in ElasticSearch (white spaces, leading _ etc.)
- [x] Ability to configure ES parameters (host & port)

### PR DESCRIPTION
**Pull request by rabbitonweb**
Fixes #1675

### CODE REVIEW

**Comment by rabbitonweb** [comment]
No longer  a WIP

**Comment by rabbitonweb** [comment]
@djspiewak this is now ready for review

File: sparkcore/src/main/scala/quasar/physical/sparkcore/fs/elastic/ElasticCall.scala [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+  
+  object Ops {
+    implicit def apply[S[_]](implicit S: ElasticCall :<: S): Ops[S] = new Ops[S]
+  }
+
+  //TODO_ES reuse single client per interpreter, close when unmounted
+  implicit def interpreter: ElasticCall ~> Task = new (ElasticCall ~> Task) {
+
+    def _listIndices: Task[List[String]] = Task.delay {
+      val client = HttpClient(ElasticsearchClientUri("localhost", 9200))
```
Comment (djspiewak) [inline sparkcore/src/main/scala/quasar/physical/sparkcore/fs/elastic/ElasticCall.scala]: None of this `HttpClient` usage is async.  You're blocking on it and side-effecting within the `Task`, which is unnecessary and eats threads.  Remember that everything in `HttpClient` is *already* going to be in `Task`, and those tasks will be properly non-blocking (or rather, blocking, but not thread-consuming).  This is a really important thing to fix.

File: sparkcore/src/main/scala/quasar/physical/sparkcore/fs/elastic/ElasticCall.scala [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+            .withBody(s"""{
+                           "source": { "index": "${src.index}", "type": "${src.typ}" }, 
+                           "dest": { "index": "${dst.index}", "type": "${dst.typ}" }
+                       }""")
+        )
+    } yield ()
+
+    def _copyIndex(src: String, dst: String): Task[Unit] = for {
+      httpClient <- Task.delay(PooledHttp1Client())
+      result          <- httpClient.expect[String](
```
Comment (djspiewak) [inline sparkcore/src/main/scala/quasar/physical/sparkcore/fs/elastic/ElasticCall.scala]: This appears to be the only async `HttpClient` usage.

File: sparkcore/src/main/scala/quasar/physical/sparkcore/fs/elastic/package.scala [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+
+    val genScWithJar: Free[S, String \/ SparkContext] = lift((for {
+      sc <- coreGenSc(sparkConf)
+      jar <- sparkCoreJar
+    } yield {
+      sc.addJar(posixCodec.printPath(jar))
+      sc
+    }).run).into[S]
+
+    val definition: SparkContext => Free[S, SparkFSDef[Eff, S]] =
```
Comment (djspiewak) [inline sparkcore/src/main/scala/quasar/physical/sparkcore/fs/elastic/package.scala]: Given that this is a brand new backend, is there a reason not to implement it on `BackendModule`?  Porting the old backends does take time, but I would really rather see all new work go under `BackendModule` at a minimum.

**Comment by djspiewak** [comment]
This also needs a merge with master, since there are conflicts.

File: sparkcore/src/main/scala/quasar/physical/sparkcore/fs/elastic/ElasticCall.scala (hunk 1 above)
Comment (rabbitonweb) [inline sparkcore/src/main/scala/quasar/physical/sparkcore/fs/elastic/ElasticCall.scala]: HttpClient is async but it uses `Future` underneath (it is not `HttpClient` from `http4s`). I tried to build a smooth transformation from `Future` to `Task` but I failed. Maybe some pair programming on it today?

File: sparkcore/src/main/scala/quasar/physical/sparkcore/fs/elastic/package.scala (hunk 3 above)
Comment (rabbitonweb) [inline sparkcore/src/main/scala/quasar/physical/sparkcore/fs/elastic/package.scala]: This is a bit more complicated. All spark based connectors are based on spark-core (*which is not a connector itself- more a foundation for writting the spark based connectors). 

The reason I want to close `spark-elastic` and `spark-cassandra` first is so that the next step will be rewritting only spark-core to `Backend module` (if I understood correctly how backend module works).

tl;dr:
1. working `spark-elastic`
2. working `s3` and `ftp` on `spark-hdfs`
3. working `spark-cassandra` 
4. rewrite `spark-core` to `backen module`

**Comment by rabbitonweb** [comment]
@djspiewak this is ready

**Comment by djspiewak** [comment]
@rabbitonweb Looks good to me on the changes front, but it doesn't actually compile.  This is ready for merge once it builds.

**Comment by rabbitonweb** [comment]
@djspiewak that's embarrassing :)

**Comment by djspiewak** [comment]
Doesn't compile

**Comment by alissapajer** [comment]
@rabbitonweb `mongodb_2_6` and `mongodb_3_0` are both still failing with 
```
java.lang.IllegalArgumentException: Argument unrecognized by ScalaTest's Runner: xonly
```

**Comment by rabbitonweb** [comment]
I try to understand why ScalaTest is being involved here. This is bizarre

**Comment by jedesah** [comment]
@rabbitonweb It's that dependency I mentioned the other day that's bringing it in.

**Comment by rabbitonweb** [comment]
@jedesah i did excluded scala test from it before but I think it was missing `_2.11` so I just fixed that, CI is running - lets see what happens

**Comment by rabbitonweb** [comment]
btw fun fact: I can not reproduce this error locally
